camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java

```diff
@@ -182,63 +257,112 @@ public Entity call() throws Exception {
                     throw new IllegalStateException("Unexpected scope "+scope);
             }
             
-            Optional<Entity> result = Iterables.tryFind(entitiesToSearch, EntityPredicates.configEqualTo(BrooklynCampConstants.PLAN_ID, componentId));
+            String desiredComponentId;
+            if (componentId == null) {
+                if (componentIdSupplier == null) {
+                    throw new IllegalArgumentException("No component-id or component-id supplier, when resolving entity in scope '" + scope + "' wrt " + entity);
+                }
+                
+                Maybe<Object> maybeComponentId = Tasks.resolving(componentIdSupplier)
+                        .as(Object.class)
+                        .context(getExecutionContext())
+                        .immediately(immediate)
```

Comment:
For the `false` case would make more sense to skip `if (timeout == null) timeout = ValueResolver.NON_BLOCKING_WAIT;` in the method.